Python/specialize.c

```diff
@@ -1813,112 +1813,110 @@ _Py_Specialize_BinarySubscr(
     cache->counter = adaptive_counter_cooldown();
 }
 
-void
-_Py_Specialize_StoreSubscr(_PyStackRef container_st, _PyStackRef sub_st, _Py_CODEUNIT *instr)
-{
-    PyObject *container = PyStackRef_AsPyObjectBorrow(container_st);
-    PyObject *sub = PyStackRef_AsPyObjectBorrow(sub_st);
 
-    assert(ENABLE_SPECIALIZATION);
-    _PyStoreSubscrCache *cache = (_PyStoreSubscrCache *)(instr + 1);
-    PyTypeObject *container_type = Py_TYPE(container);
-    if (container_type == &PyList_Type) {
-        if (PyLong_CheckExact(sub)) {
-            if (_PyLong_IsNonNegativeCompact((PyLongObject *)sub)
-                && ((PyLongObject *)sub)->long_value.ob_digit[0] < (size_t)PyList_GET_SIZE(container))
-            {
-                instr->op.code = STORE_SUBSCR_LIST_INT;
-                goto success;
-            }
-            else {
-                SPECIALIZATION_FAIL(STORE_SUBSCR, SPEC_FAIL_OUT_OF_RANGE);
-                goto fail;
-            }
-        }
-        else if (PySlice_Check(sub)) {
-            SPECIALIZATION_FAIL(STORE_SUBSCR, SPEC_FAIL_SUBSCR_LIST_SLICE);
-            goto fail;
-        }
-        else {
-            SPECIALIZATION_FAIL(STORE_SUBSCR, SPEC_FAIL_OTHER);
-            goto fail;
-        }
-    }
-    if (container_type == &PyDict_Type) {
-        instr->op.code = STORE_SUBSCR_DICT;
-        goto success;
-    }
 #ifdef Py_STATS
+static int
+store_subscr_fail_kind(PyObject *container_type)
+{
     PyMappingMethods *as_mapping = container_type->tp_as_mapping;
     if (as_mapping && (as_mapping->mp_ass_subscript
                        == PyDict_Type.tp_as_mapping->mp_ass_subscript)) {
```

Comment:
I *think* it's possible for a data race to occur between the load in `as_mapping->mp_ass_subscript` and an assignment to the same field in another thread (potentially as a result of Python code assigning to the type's  `__setitem__` attr). I'm not totally sure what we want to do here, but a couple of thoughts:

1. We could use an atomic load in `as_mapping->mp_ass_subscript` and change the code in `typeobject.c` that updates the slots to use atomic stores. 
2. Or, since this only happens in stats builds and is benign (to the extent that a data race is benign), we could just ignore it.

Reply:
I think the there's a general risk of a race with reassignment to `mp_ass_subscript` that's not specific to specialization -- it's true for every access of `mp_ass_subscript`, including in `PyObject_SetItem`, and probably true for the type slots in general.

I'm not exactly sure how we should address them. I'd lean towards doing a stop-the-world pause when modifying the slot so that we don't need to change every read. I expect that to be a bit tricky since class definitions look like they're mutating the class.

Reply:
Oof ok, I assumed we were already handling this for the non-specialization cases. Definitely not something to be solved in this PR. I didn't find an existing issue tracking this, so I filed https://github.com/python/cpython/issues/127266.